### ISSUE TITLE
Question: Is it possible to provide external custom operators and comparators without forking the library?

### ISSUE THREAD

**Issue by simonepri** [facebookresearch/PyTorch-BigGraph#51]

**Comment by lerks** [comment]
Yes, although it cannot be done just by editing the config, you will have to write a new training entry point, i.e., a new top-level script.

For example, suppose you want to introduce a L2 norm comparator (this code is for example only, it is not designed to be efficient or even correct, L2 norm should be used with caution because, contrary to the dot product or cosine distance, it cannot be expressed as a matrix multiply and thus cannot be accelerated). You could do it as follows:
```py
import torch

from torchbiggraph.config import parse_config
from torchbiggraph.model import instantiate_operator, AbstractComparator, MultiRelationEmbedder
from torchbiggraph.train import train
from torchbiggraph.types import Side


class L2Comparator(AbstractComparator):

    def prepare(self, embs):
        return embs

    def forward(self, lhs_pos, rhs_pos, lhs_neg, rhs_neg):
        return (
            torch.dist(lhs_pos - rhs_pos),
            torch.dist(lhs_neg.unsqueeze(-3) - rhs_pos.unsqueeze(-2)),
            torch.dist(lhs_pos.unsqueeze(-2) - rhs_neg.unsqueeze(-3)),
        )


def main():
    # Obtain some config, with any value of comparator.
    config = parse_config("config.py")

    # The following is almost a copy of the make_model function for non-dynamic relations.
    model = MultiRelationEmbedder(
        config.dimension,
        config.relations,
        config.entities,
        num_uniform_negs=config.num_uniform_negs,
        num_batch_negs=config.num_batch_negs,
        lhs_operators=[instantiate_operator(r.operator, Side.LHS, 0, config.dimension)) for r in config.relations],
        rhs_operators=[instantiate_operator(r.operator, Side.RHS, 0, config.dimension)) for r in config.relations],
        comparator=L2Comparator(),
        global_emb=config.global_emb,
        max_norm=config.max_norm,
    )

    train(config, model=model)


if __name__ == "__main__":
    main()
```
(Disclaimer: I did not test this code!)

A similar approach would work for operators. In the same way, you can provide a custom subclass of `Trainer` to the `train` function (using the `trainer` argument) which allows you to change how aspects like the batching or the loss function work.

Does this answer your question?

**Comment by simonepri** [comment]
Thank you for the code Luca.

It that looks far from being easy to use at the moment.
From a user point-of-view I would expect more something like this:

```python
import torch

from torchbiggraph.config import ConfigSchema
from torchbiggraph.model import AbstractComparator

class L2Comparator(AbstractComparator):
    def prepare(self, embs):
        return embs
    def forward(self, lhs_pos, rhs_pos, lhs_neg, rhs_neg):
        return (
            torch.dist(lhs_pos - rhs_pos),
            torch.dist(lhs_neg.unsqueeze(-3) - rhs_pos.unsqueeze(-2)),
            torch.dist(lhs_pos.unsqueeze(-2) - rhs_neg.unsqueeze(-3)),
        )

# Load configs
entity_name = "e"
config_dict = {
    # Other configs
    "operator": "l2norm",
}
config = ConfigSchema.from_dict(config_dict, custom_comparators={'l2norm': L2Comparator})

# Train
train(config, custom_comparators={'l2norm': L2Comparator})
```

PS: I believe that a correct implementation of L2 and L1 would be really helpful to have in the framework.

**Comment by lerks** [comment]
So, first of all, if I remember correctly, one reason we haven't added L1 and L2 norms is that we haven't found them to be superior to the dot product and the cosine distance, even though they are sensibly less efficient. By not providing them, we're hoping to nudge people to other methods and discourage and curb their usage.

As for how to expose entry points for custom plug-ins, well, my best answer is that API design is hard and never perfect for everyone. A concern we kept in mind while writing the current way of doing it is that we wanted to allow people to plug in _anything_ that vaguely resembled a model, and not just something based off MultiRelationalEmbedder with its rigid operator+comparator structure. This can definitely be improved, possibly by extending the `make_model` function (what if it accepted a `comparator=` argument which would override the config?) or providing new helper functions. However your proposed approach, if adopted exclusively, wouldn't satisfy the above requirement.

Re-factoring the model and training loop for better "extendibility" is something we're still actively working on so if you have more proposals or want to refine the above one we're very interested!

**Comment by simonepri** [comment]
> By not providing them, we're hoping to nudge people to other methods and discourage and curb their usage.

I understand that practitioner should avoid them, but from an academic perspective it may be interesting to have them.

> As for how to expose entry points for custom plug-ins, well, my best answer is that API design is hard and never perfect for everyone.

I agree, I was just trying to explain what I would like to have had in the (I believe common) situation in which one wants to provide only a custom `operator` and a custom `comparator`.

These were the two main points I had in my mind when I provided the code above:
* `ConfigSchema.from_dict` should somehow allow me to define whether some values I define as `operator` or `comparator` should be considered valid even though they are not one of the built-in operators/comparators.

* `make_model` should allow me to provide the implementation for the custom operators/comparators I used in the config eventually throwing an error if I do not provide the implementation for some of them.

**Comment by simonepri** [comment]
For the sake of completeness, below there's a working L2 norm comparator (where I return the negative distance as you suggested):
It's about 20/30 times slower than the `cos` comparator on my setup.

```python
class L2Comparator(AbstractComparator):

    def prepare(
        self,
        embs: FloatTensorType,
    ) -> FloatTensorType:
        return embs

    def forward(
        self,
        lhs_pos: FloatTensorType,
        rhs_pos: FloatTensorType,
        lhs_neg: FloatTensorType,
        rhs_neg: FloatTensorType,
    ) -> Tuple[FloatTensorType, FloatTensorType, FloatTensorType]:
        num_chunks, num_pos_per_chunk, dim = match_shape(lhs_pos, -1, -1, -1)
        match_shape(rhs_pos, num_chunks, num_pos_per_chunk, dim)
        match_shape(lhs_neg, num_chunks, -1, dim)
        match_shape(rhs_neg, num_chunks, -1, dim)

        pos_scores = torch.norm(lhs_pos - rhs_pos, 2, -1).neg()
        lhs_neg_scores = torch.norm(lhs_neg.unsqueeze(-3) - rhs_pos.unsqueeze(-2), 2, -1).neg()
        rhs_neg_scores = torch.norm(lhs_pos.unsqueeze(-2) - rhs_neg.unsqueeze(-3), 2, -1).neg()

        return pos_scores, lhs_neg_scores, rhs_neg_scores
```

But it becomes just 10/15 times slower if one computes the squared l2 norm instead:
```python
        pos_scores = (lhs_pos - rhs_pos).pow(2).sum(-1).neg()
        lhs_neg_scores = (lhs_neg.unsqueeze(-3) - rhs_pos.unsqueeze(-2)).pow(2).sum(-1).neg()
        rhs_neg_scores = (lhs_pos.unsqueeze(-2) - rhs_neg.unsqueeze(-3)).pow(2).sum(-1).neg()
```

**Comment by lerks** [comment]
Thanks for your investigation into an L2 comparator, I am sure it will help quite a few people!

Have you tried other ways of converting a distance into a score other than taking the negative? For example, taking the inverse? And did you check what impact on model quality these different methods have? Also, how does this change between "regular" L2 norm and squared L2 norm (i.e., just sum of squares)?

Also, we don't recommend to use `all_negs` during training, as it scales very poorly (as you will probably have realized). It's there as it's occasionally helpful in evaluating small datasets. It's a bit "more ok" to use it during evaluation as one can do smaller batches without adverse effects in case one needs to reduce the memory usage.

Unless there is a good reason for PyTorch's cdist from being undocumented, you may want to open an issue on their repo to report this, maybe they just forgot to add it to the docs.

**Comment by adamlerer** [comment]
Wow, I didn't know about `cdist`, great find @simonepri . 

@lerks using the negative distance as the score is the typical approach (e.g. see the TransE paper https://www.aclweb.org/anthology/D15-1034 ) , not something like 1/dist . It's possible that some other metric would work but if we're looking to match the standard approaches we should stick with negative distance.

**Comment by simonepri** [comment]
I realized that the last implementation I provided is flawed.

In particular when I was doing this:
```python
rhs_neg_scores = torch.cdist(
  lhs_pos.view(-1, dim),
  rhs_neg.view(-1, dim),
  p=self.p,
).view(num_chunks, num_pos_per_chunk, -1).neg()
# rhs_neg_scores here has size C x P x C x L (which is wrong)
```
It actually needed to be like this:
```python
rhs_neg_scores = torch.empty(lhs_pos.size(0), lhs_pos.size(1), rhs_neg.size(1))
for i in range(0, lhs_pos.size(0)):
    rhs_neg_scores[i] = torch.cdist(lhs_pos[i], rhs_neg[i], p=self.p).neg()
# rhs_neg_scores here has size C x P x L (which is correct)
```

I then spent some time to find a better and faster solution, and I came out with the following code:
```python
import time
import torch

@torch.jit.script
def fast_batched_cdist_l2(x1, x2):
    x1_norm = x1.pow(2).sum(dim=-1, keepdim=True)
    x2_norm = x2.pow(2).sum(dim=-1, keepdim=True)
    res = torch.baddbmm(x2_norm.transpose(-2, -1), x1, x2.transpose(-2, -1), beta=1, alpha=-2).add_(x1_norm).sqrt_()
    return res

@torch.jit.script
def slow_batched_cdist_l2(x1, x2):
  res = torch.empty(x1.size(0), x1.size(1), x2.size(1)).cuda()
  for i in range(x1.size(0)):
    res[i] = torch.cdist(x1[i], x2[i], p=2.0)
  return res

a = torch.randn(20, 500, 400).cuda()
b = torch.randn(20, 10500, 400).cuda()

for i in range(5):
    start_time = time.time()
    res = slow_batched_cdist_l2(a, b)
    torch.cuda.synchronize()
    print(f'slow batched cdist l2 time {i}: {time.time() - start_time:.2f}s')

for i in range(5):
    start_time = time.time()
    res2 = fast_batched_cdist_l2(a, b)
    torch.cuda.synchronize()
    print(f'fast batched cdist l2 time {i}: {time.time() - start_time:.2f}s')


equal = torch.all(torch.lt(torch.abs(torch.add(res, -res2)), 1e-4))
print("Equals", int(equal) == 1)
print("---\n")
```
<img src="https://user-images.githubusercontent.com/3505087/59557707-dfb2e400-8fe0-11e9-8372-f12917b9d1e0.png" width="200">

Which is as fast as the `cos` comparator.
I will post more detailed tests tomorrow. 😴

**Comment by simonepri** [comment]
@lerks I feel that we definitely went OT in this issue and it was partially my fault.

The main takeaways of this issue are:
- we should improve [the documentation](https://user-images.githubusercontent.com/3505087/59562860-3b598d80-9032-11e9-9b72-6c4dd38e87d1.png) about custom operators and comparators.
- user may like a more "plug and play" way to provide custom operators and comparators.
- the `l2` norm can be implemented to be as fast as the `cos` and `dot` comparators

**Comment by lerks** [comment]
@simonepri What do you think of the approach I just pushed?
